### PR TITLE
[PluggableDevice] Enable TF_TensorFromProto

### PR DESCRIPTION
**Pull request by Solaryee**
Add a new C API `TF_TensorFromProto`, which is designed for converting TensorProto(TF_Buffer) to Tensor(TF_Tensor). Typical use case is to get tensor value from TensorProto in graph pass.

### CODE REVIEW

File: tensorflow/c/c_api.cc [hunk 1]
```diff
@@ -154,6 +155,20 @@ void TF_DeleteBuffer(TF_Buffer* buffer) {
 
 TF_Buffer TF_GetBuffer(TF_Buffer* buffer) { return *buffer; }
 
+void TF_TensorFromProto(const TF_Buffer* from, TF_Tensor* to,
+                        TF_Status* status) {
+  TF_SetStatus(status, TF_OK, "");
+  tensorflow::TensorProto from_tensor_proto;
+  Status cc_status;
```
Comment (saxenasaurabh) [inline tensorflow/c/c_api.cc:162]: We could avoid creating `cc_status` here I think:

```
status->status = BufferToMessage(from, &from_tensor_proto);
if (!status->status.ok()) {
  return;
} 
status->status = tensorflow::down_cast<tensorflow::TensorInterface*>(to->tensor)->FromProto(from_tensor_proto);
```

**Review by saxenasaurabh — CHANGES_REQUESTED** [review]

File: tensorflow/c/c_api.cc (hunk 1 above)
Comment (ShengYang1) [reply tensorflow/c/c_api.cc]: Yes, that makes sense. Done.

**Review by saxenasaurabh — APPROVED** [review]

File: tensorflow/c/c_api_test.cc [hunk 2]
```diff
@@ -1543,6 +1543,33 @@ TEST(CAPI, TestBitcastFrom_Reshape) {
...
+  const int num_bytes = 6 * sizeof(float);
+  float* values =
+      reinterpret_cast<float*>(tensorflow::cpu_allocator()->AllocateRaw(
+          EIGEN_MAX_ALIGN_BYTES, num_bytes));
+  int64_t dims[] = {2, 3};
+  bool deallocator_called = false;
+  TF_Tensor* t_c = TF_NewTensor(TF_FLOAT, dims, 2, values, num_bytes,
+                                &Deallocator, &deallocator_called);
+
+  TF_Status* status = TF_NewStatus();
```
Comment (saxenasaurabh) [inline tensorflow/c/c_api_test.cc:1564]: This is leaking. Please add `TF_DeleteStatus(status);`

File: tensorflow/c/c_api_test.cc [hunk 3]
```diff
@@ -1543,6 +1543,33 @@ TEST(CAPI, TestBitcastFrom_Reshape) {
   TF_DeleteTensor(b);
 }
 
+TEST(CAPI, TestFromProto) {
+  Tensor t_cc(DT_FLOAT, TensorShape({2, 3}));
+  t_cc.flat<float>().setConstant(1.0);
+  tensorflow::TensorProto t_proto;
```
Comment (saxenasaurabh) [inline tensorflow/c/c_api_test.cc:1549]: This needs an additional include:

`#include "third_party/tensorflow/core/framework/tensor.proto.h"`

File: tensorflow/c/c_api_test.cc [hunk 4]
```diff
@@ -1543,6 +1543,33 @@ TEST(CAPI, TestBitcastFrom_Reshape) {
...
+  t_cc.AsProtoField(&t_proto);
+
+  TF_Buffer* t_buffer = TF_NewBuffer();
+  tensorflow::MessageToBuffer(t_proto, t_buffer);
```
Comment (saxenasaurabh) [inline tensorflow/c/c_api_test.cc:1553]: This returns a `Status` which cannot be ignored. Please change to:

```
  Status s = tensorflow::MessageToBuffer(t_proto, t_buffer);
  EXPECT_TRUE(s.ok()) << s.error_message();
```

**Comment by saxenasaurabh** [comment]
This is failing some tests internally. Suggested some changes that should make this work. PTAL.

File: tensorflow/c/c_api_test.cc (hunk 2 above)
Comment (ShengYang1) [reply tensorflow/c/c_api_test.cc:1565]: Done

File: tensorflow/c/c_api_test.cc (hunk 3 above)
Comment (ShengYang1) [reply tensorflow/c/c_api_test.cc:1550]: Done. And I guess what you actually mean is `#include "tensorflow/core/framework/tensor.pb.h"`

**Comment by ShengYang1** [comment]
Thanks @saxenasaurabh. Already addressed all your comments.

File: tensorflow/c/c_api_test.cc (hunk 4 above)
Comment (ShengYang1) [reply tensorflow/c/c_api_test.cc]: Done

**Review by saxenasaurabh — APPROVED** [review]

**Comment by saxenasaurabh** [comment]
> Thanks @saxenasaurabh. Already addressed all your comments.

Thanks!

**Review by saxenasaurabh — APPROVED** [review]

**Pull request merged**